dao-contracts/src/onboarding_voter.rs

```diff
@@ -0,0 +1,213 @@
+    fn get_kyc_token_address(&self) -> Address;
+    fn get_va_token_address(&self) -> Address;
+
+    fn get_voting(&self, voting_id: U256) -> Option<Voting>;
+    fn get_ballot(&self, voting_id: U256, address: Address) -> Option<Ballot>;
+    fn get_voter(&self, voting_id: U256, at: u32) -> Option<Address>;
+}
+
+#[derive(Instance)]
+pub struct OnboardingVoterContract {
+    onboarding: OnboardingInfo,
+    kyc: KycInfo,
+    voting: GovernanceVoting,
+    sequence: SequenceGenerator,
+}
+
+impl OnboardingVoterContractInterface for OnboardingVoterContract {
+    fn init(
+        &mut self,
+        variable_repo: Address,
+        reputation_token: Address,
+        kyc_token: Address,
+        va_token: Address,
+    ) {
+        self.onboarding.init(va_token);
+        self.kyc.init(kyc_token);
+        self.voting.init(variable_repo, reputation_token);
+    }
+
+    delegate! {
+        to self.onboarding {
+            fn get_va_token_address(&self) -> Address;
+        }
+
+        to self.kyc {
+            fn get_kyc_token_address(&self) -> Address;
+        }
+
+        to self.voting {
+            fn get_variable_repo_address(&self) -> Address;
+            fn get_reputation_token_address(&self) -> Address;
+            fn get_dust_amount(&self) -> U256;
+            fn get_voting(&self, voting_id: U256) -> Option<Voting>;
+            fn get_ballot(&self, voting_id: U256, address: Address) -> Option<Ballot>;
+            fn get_voter(&self, voting_id: U256, at: u32) -> Option<Address>;
+        }
+    }
+
+    fn create_voting(&mut self, action: onboarding::Action, subject_address: Address, stake: U256) {
+        self.assert_no_ongoing_voting(&subject_address);
+
+        let (entry_point, runtime_args) = match action {
+            onboarding::Action::Add => self.configure_add_voting(subject_address),
+            onboarding::Action::Remove => self.configure_remove_voting(subject_address),
+        };
+        let creator = caller();
+        let contract_to_call = self.get_va_token_address();
+
+        self.voting
+            .create_voting(creator, stake, contract_to_call, entry_point, runtime_args);
+        self.onboarding.set_voting(&subject_address);
+    }
+
+    fn vote(&mut self, voting_id: VotingId, choice: Choice, stake: U256) {
+        let voter = caller();
+        self.voting.vote(voter, voting_id, choice, stake);
+    }
+
+    fn finish_voting(&mut self, voting_id: VotingId) {
+        let address = self.extract_address_from_args(voting_id);
+        self.voting.finish_voting(voting_id);
+        self.onboarding.clear_voting(&address);
+    }
+}
+
+// non-contract implementation
+impl OnboardingVoterContract {
+    fn configure_remove_voting(&mut self, subject_address: Address) -> (String, RuntimeArgs) {
+        self.assert_onboarded(&subject_address);
+        self.assert_has_reputation(&subject_address);
+
+        let token_id = self.onboarding.token_id_of(&subject_address).unwrap();
```

Comment:
or_revert_with?

Reply:
Ok, I moved error handling to the module, so it returns already unwrapped token_id or reverts with a proper error